File: problems/medium/preorder_binary_tree.rb
# https://leetcode.com/problems/binary-tree-preorder-traversal/submissions/
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def preorder_traversal(root)
  @result = []
  traverse(root)
  @result
end

def traverse(root)
 return if root.nil?
 @result.push(root.val)
 traverse(root.left)
 traverse(root.right)
end

File: problems/medium/preorder_binary_tree.go
// https://leetcode.com/problems/binary-tree-preorder-traversal/
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func preorderTraversal(root *TreeNode) []int {
  result := []int
  result = traverse(root, result)
  return result
}

func traverse(root *TreeNode, result []int) []int {
  if root == nil {
    return result
  }
  result = append(result, root.Val)
  result = traverse(root.Left, result)
  result = traverse(root.Right, result)
  return result
}